### ISSUE TITLE
[RTDB] Wrong "Non-default "firebase-admin" instance created!" message

### ISSUE THREAD

**Issue by ftzi** [firebase/firebase-tools#2670]
### [REQUIRED] Environment info

**firebase-tools:** 8.11.2

**Platform:** Ubuntu 20.04

### [REQUIRED] Test case

Unfortunately I am in a rush to finish my app and I can't fill this issue correctly, and my project is currently fairly big and close to the prod to be safely published here. As the issue pointed here isn't a big deal and won't halt my development, I really won't invest more time on it.

### [REQUIRED] Steps to reproduce

1) Have an emulator that uses more than one Realtime Database instances that are correctly set under .firebaserc:
```
{
  "projects": {
    "default": myProjectName
  },
  "targets": {
    myProjectName: {
      "database": {
        "instances": [
           "myInstance-0", // (not real name)
           ...1 etc
```

2) Run the emulator, and all the instances appear there correctly (Also appear an instance named just "0", but I may open an issue for that later).

3) Have in your code the instance initialization
```
const dbs[dbId] = firebase.database(firebase.initializeApp({
    databaseURL: `https://myInstance-${dbId}.firebaseio.com/`
  }, dbId)));
```

4) The warning will pop:

```
Non-default "firebase-admin" instance created!
[1]    - This instance will *not* be mocked and will access production resources.
```

5) However, when setting any data to this Realtime Database instance, it will ONLY appear in the emulator UI as it should happen, and my prod DB is intact. So, this warn is NOT true and shouldn't appear.

### [REQUIRED] Expected behavior

This warn shouldn't appear if the data is going to the emulator DB.

### [REQUIRED] Actual behavior

This warn shouldn't appear if the data is going to the emulator DB.

Sorry for not filling correctly this Issue.

**Comment by samtstern** [comment]
@SrBrahma thanks!  It does look like we need to fix those logs.

**Comment by Agusfn** [comment]
@samstern I have a somewhat related problem, but related with the excecution instancing. 

My cloud functions use two databases from the same proyect, one of them is of course the default. When I initialize the second one (running with the emulator), I get the warning `Non-default "firebase-admin" instance created!` as well, but that's not the problem.

I do the second db/app instancing inside a class, to do it when needed only, and not in index.js (I am not sure if this is proper, but I wouldn't know how to access that variable from other scope). 
The problem I am experiencing is that on calling the second initializeApp, i think the emulator creates a new excecution instance without running index.js again. The thing is that on index.js i've set up some important config initialization, and the rest of the excecution takes place without the config, resulting most likely in an error because lack of config.

This didn't happen to me previously. I've upgraded several packages and implemented typescript so it might have been after one of those upgrades.

**Comment by SrBrahma** [comment]
> @samstern I'm experiencing that problem, but related with the excecution instancing. 
> 
> My cloud functions use two databases from the same proyect, one of them is of course the default. When I initialize the second one (running with the emulator), I get the warning `Non-default "firebase-admin" instance created!` as well, but that's not the problem.
> 
> I do the second db/app instancing inside a class, to do it when needed only, and not in index.js (I am not sure if this is proper, but I wouldn't know how to access that variable from other scope). 
> The problem I am experiencing is that on calling the second initializeApp, i think the emulator creates a new excecution instance without running index.js again. The thing is that on index.js i've set up some important config initialization, and the rest of the excecution takes place without the config, resulting most likely in an error because lack of config.
> 
> This didn't happen to me previously. I've upgraded several packages and implemented typescript so it might have been after one of those upgrades.

I have 10 dbs in my project. I have a firebase.ts (can be js) that have in a object, for each dbId as key, its initApp as value. The entry is only created when it's used and the entry doesn't exists. I use that for both client and cloud functions.

If you want, I can share my code here / create a npm package.

**Comment by Agusfn** [comment]
@samstern You still have the warning right? Do you make any kind of custom initialization in index.js? Because i believe my problem is being cause for that.

**Comment by Agusfn** [comment]
@SrBrahma You still have the warning right? Do you have an custom initialization on index.ts (or js)? Because I believe my problem is being cause because of that.

**Comment by SrBrahma** [comment]
> @SrBrahma You still have the warning right? Do you have any custom initialization on index.ts (or js)? Because I believe my problem is being cause because of that.

Yup, the warning happens but you can ignore it, it's just a "visual bug".

My index doesn't have any firebase configs, as I use the firebase file as I mentioned. When I access my pc I will post here my relevant code.

**Comment by Agusfn** [comment]
To be more specific, I have a `Config `class that has an `initialize()` static method, it loads all the remote config to be able to use it from any function without having to initialize it in each function. When a new db/app instance is made, it skips that Config initialization. I don't know if index.js is safe to do that thing

**Comment by Agusfn** [comment]
@SrBrahma Okay aparently my excecution environment problem is not related to the initialization of the second database, so my problem is not related with this issue, I should have tested a bit more before commenting. 

The problem arises when trying to set a static variable of a class from the index.js file ([called at every cold-start](https://cloud.google.com/functions/docs/concepts/exec#function_scope_versus_global_scope)) and assuming that will remain in the scope or the excecution instance of the functions. This was true for me prior to updating my functions, but something has changed.

 I should make a different issue for this since this related with the function scope and global scope.

**Comment by SrBrahma** [comment]
<details>
<summary>My firebase.ts</summary>

```ts

import * as functions from 'firebase-functions';
import * as admin from 'firebase-admin';


const firebaseApp = admin.initializeApp();


export { functions };

export const auth = admin.auth();

export { firebaseApp as firebase };

export const emulatingFunctions = !!process.env.FUNCTIONS_EMULATOR;
export const emulatingDatabase = !!process.env.FIREBASE_DATABASE_EMULATOR_HOST;
export const emulatingAuth = !!process.env.FIREBASE_AUTH_EMULATOR_HOST;
export const emulatingAll = emulatingFunctions && emulatingDatabase && emulatingAuth;

// https://stackoverflow.com/a/63127747/10247962

```

</details>


<details>
<summary>My database.ts</summary>

```ts
import * as admin from 'firebase-admin';
import { HandlerF } from 'firebase-functions-extended';


export type Database = ReturnType<(typeof admin.database)>;


export const masterDb = admin.database();

/** As any so it can be used in models. Actually a function. */
export const ServerIncrement = admin.database.ServerValue.increment as any;
/** As any so it can be used in models. Actually an object. */
export const ServerTimestamp = admin.database.ServerValue.TIMESTAMP as any;

const dbs: { [x: string]: Database; } = {};


// Includes this Id
export const maxDbId = 9;


// Later may add ([A-Z]?) to allow special dbs
// You still have to check after this if the id is <= maxDbId.
// Won't allow 0 in first digit if have another digit after.
const dbIdRegex = (/^(([1-9][0-9]{0,2})|0)$/);


// dbId is 'yyy' or 'Xyyy', being y numbers and x a possible special uppercase char.
/**
 * Returns the Db or throw an error if error.
 *
 * validateId defaults to true
 * */
export function getDb(dbId: string, validateId: boolean = true): Database {
  // Check if the db is already stored.
  if (dbs[dbId])
    return dbs[dbId];

  if (validateId)
    if (!dbIdRegex.test(dbId) || Number(dbId) > maxDbId)
      throw new Error('Identificador do banco de dados inválido.');

  return (dbs[dbId] = admin.database(admin.initializeApp({
    databaseURL: `https://***-${dbId}.firebaseio.com/` // *** is my db base name, hid in this issue
  }, dbId)));
}





/** Will return the DB using the dbId */
export const auxGetDb: HandlerF<{ dbId: string; }, {}, { db: Database; }> = (
  { data: { dbId }, ExtError }
) => {
  try {
    return { db: getDb(dbId) };
  } catch (err) {
    throw ExtError(err.message);
  }
};

// Don't mind that HandlerF, it is an intermediary between another stuff in my code.

```

</details>

**Comment by SrBrahma** [comment]
The consts (and maybe variables? dont remember) may be shared between Cloud Functions instances. This will only create the db reference if it doesn't already exists.

Idk if this really increases the performance (as in my case I have 10 DB instances). For your case, 2 DBs, you could have them always created. Follow this code idea and adapt to your needs.

**Comment by Agusfn** [comment]
@SrBrahma I really appreciate you for taking your time to prepare and show me your code! It is indeed a nice way to instanciate several databases and I will take it into consideration for when I need to implement more.

However, my problem ended up being totally different and doesn't have to do with databases, lol. So that was a bad thing from my part. I will make a separate thread in SO or github issue to check it. Thanks!
